### DISCUSSION TITLE
Accessing a model using with_polymorphic has name of None

### DISCUSSION THREAD

**Discussion by ske99ey** [sqlalchemy/sqlalchemy#11326] [Usage Questions] [answered]
Hi, I am working on updating a code base from 1.4 to 2.X and am trying to clear up a warning in regards to `Query.with_polymorphic()`. It is still using the query syntax which I'd rather not update at this time.

Generally speaking we have the following models:
```
class Fruit(Base):
  __mapper_args__ = {"polymorphic_on": type}

class Orange(Fruit):
  __mapper_args__ = {"polymorphic_identity": FruitEnum.Orange}

class Apple(Fruit):
  __mapper_args__ = {"polymorphic_identity": FruitEnum.Apple}

class Fuji(Apple):
  __mapper_args__ = {"polymorphic_identity": FruitEnum.Fuji}

class Mcintosh(Apple):
  __mapper_args__ = {"polymorphic_identity": FruitEnum.Mcintosh}
```

The warning is coming from the following since its [no longer supported in 2+](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.with_polymorphic):
```
query = models.Fruit.query.with_polymorphic("*")
```

I attempted to rewrite it as documented with `with_polymorphic()`:
```
FruitWithPolymorphic = with_polymorphic(models.Fruit, "*")
query = db.session.query(FruitWithPolymorphic)
```

But am encountering an issue with using the results from the query as it doesn't have the Field of the Polymorphic class defined (in this case Fruit):
> AttributeError: Could not locate column in row for column 'Fruit'

Inspecting the query reveals that the polymorphic aliased class is not getting a name and thus its set to None causing it to not be defined as a field. If I add `name` to the [construction](https://github.com/sqlalchemy/sqlalchemy/blob/e15890b455b14542bb17cdb208e6d3b723799ed6/lib/sqlalchemy/orm/util.py#L1089-L1097) of the `AliasedClass` within `with_polymorphic()` then everything works.

Am I doing something wrong or is there an issue with `with_polymorphic()`?

**Comment by ske99ey** [comment]
@zzzeek, sorry to ask directly but would you be willing to weigh in on this?

**Comment by zzzeek** [comment]
hi -

I need a runnable example to see why that error would occur.    Since we missed this Q, here's a runnable example, modify it to show your error (no dependencies please):

```py
from enum import Enum

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Enum as Enum_typ
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import query
from sqlalchemy.orm import Session
from sqlalchemy.orm import with_polymorphic

Base = declarative_base()


class FruitEnum(Enum):
  Orange = "orange"
  Apple = "apple"
  Fuji = "Fuji"
  Mcintosh = "Mcintosh"

class Fruit(Base):
    __tablename__ = "fruit"
    id = Column(Integer, primary_key=True)
    type = Column(Enum_typ(FruitEnum))
    __mapper_args__ = {"polymorphic_on": type}

class Orange(Fruit):
    __tablename__ = "orange"
    __mapper_args__ = {"polymorphic_identity": FruitEnum.Orange}
    id = Column(Integer, ForeignKey('fruit.id'), primary_key=True)

class Apple(Fruit):
    __tablename__ = "apple"
    __mapper_args__ = {"polymorphic_identity": FruitEnum.Apple}
    id = Column(Integer, ForeignKey('fruit.id'), primary_key=True)

class Fuji(Apple):
    __tablename__ = "fuji"
    __mapper_args__ = {"polymorphic_identity": FruitEnum.Fuji}
    id = Column(Integer, ForeignKey('apple.id'), primary_key=True)

class Mcintosh(Apple):
    __tablename__ = "mcintosh"
    __mapper_args__ = {"polymorphic_identity": FruitEnum.Mcintosh}
    id = Column(Integer, ForeignKey('apple.id'), primary_key=True)



e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add_all([Mcintosh(), Orange(), Apple()])
s.commit()


s = Session(e)
FruitWithPolymorphic = with_polymorphic(Fruit, "*")
query = s.query(FruitWithPolymorphic)
print(query.all())
```

**Reply by zzzeek** [reply]
the reason we just need such an example is extremely often, I'm not able to decipher the descriptions of these problems, which often points to some kind of misunderstanding that's actually the issue:

> But am encountering an issue with using the results from the query as it doesn't have the Field of the Polymorphic class defined (in this case Fruit):

Your mapping seems to indicate the name of your polymorphic field is "type".  the word "Fruit" seems like it would be the value stored in a column.   Your error message seems to indicate a column named "Fruit" is not found, which seems unusual, why is the field called "type" when you actually called your column "Fruit" ?  A capitalized name like "Fruit" for a column name here also seems unusual.    It's also not a normal thing that SQLAlchemy omits the polymorphic column, that's never been an issue ever so it seems like something else is off that would produce this result (I'd have to build an extremely weird and awkward example to make it produce that error).


> Inspecting the query reveals that the polymorphic aliased class is not getting a name 

that also is not a sentence I understand.  SQL queries don't have polymorphic aliased classes inside them, there might be a subquery, and if there were here, it definitely would have an anonymized name - the name does not matter.   But in this case there is no named subquery anyway, in modern SQLAlchemy versions (1.4 and 2.0 certainly) the query is a flat join:

```
SELECT fruit.id AS fruit_id, fruit.type AS fruit_type, orange.id AS orange_id, apple.id AS apple_id, fuji.id AS fuji_id, mcintosh.id AS mcintosh_id 
FROM fruit LEFT OUTER JOIN orange ON fruit.id = orange.id LEFT OUTER JOIN apple ON fruit.id = apple.id LEFT OUTER JOIN fuji ON apple.id = fuji.id LEFT OUTER JOIN mcintosh ON apple.id = mcintosh.id

```

**Comment by ske99ey** [comment]
Thanks for the initial outline - we were able to modify it to demonstrate our issue. See the comments near the end for where we see our issue and as mentioned initially where using `query.with_polymorphic` worked.

```py
from enum import Enum

from sqlalchemy import Column, create_engine, Enum as Enum_typ, ForeignKey, Integer
from sqlalchemy.orm import (
    declarative_base,
    relationship,
    Session,
    with_polymorphic,
)

Base = declarative_base()


class Owner(Base):
    __tablename__ = "owner"
    id = Column(Integer, primary_key=True)


class FruitEnum(Enum):
    Orange = "orange"
    Fuji = "Fuji"
    Mcintosh = "Mcintosh"


class Fruit(Base):
    __tablename__ = "fruit"
    id = Column(Integer, primary_key=True)
    type = Column(Enum_typ(FruitEnum))
    owner_id = Column(Integer, ForeignKey(Owner.id))
    owner = relationship(Owner)
    __mapper_args__ = {"polymorphic_on": type}


class Orange(Fruit):
    __mapper_args__ = {"polymorphic_identity": FruitEnum.Orange}


class Apple(Fruit):
    __abstract__ = True


class Fuji(Apple):
    __mapper_args__ = {"polymorphic_identity": FruitEnum.Fuji}


class Mcintosh(Apple):
    __mapper_args__ = {"polymorphic_identity": FruitEnum.Mcintosh}


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

owner = Owner()

mcintosh = Mcintosh()
mcintosh.owner = owner

orange = Orange()
orange.owner = owner

fuji = Fuji()
fuji.owner = owner

s.add_all([owner])
s.add_all([mcintosh, orange, fuji])
s.commit()


s = Session(e)
FruitWithPolymorphic = with_polymorphic(Fruit, "*")
testQuery = s.query(FruitWithPolymorphic)
testQuery = testQuery.add_entity(Owner)

print(testQuery.first())

first = testQuery.first()

# Why does this throw
# Could not locate column in row for column 'Fruit'
try:
    print(first.Fruit)
except Exception as e:
    print("Why does first.Fruit throw an exception?")
    print(e)

# But this works
print(first.Owner)
```

**Reply by CaselIT** [reply]
the with_polymorphic construct returns a unnamed object, so you can only access it in a positional way, as `first[0]`.
To give it a name you can use aliased:
```py
s = Session(e)
FruitWithPolymorphic = with_polymorphic(Fruit, "*")
WithName = aliased(FruitWithPolymorphic, name='Fruit')
testQuery = s.query(WithName)
testQuery = testQuery.add_entity(Owner)

for r in testQuery.all():
    print(r.Fruit)
    print(r.Owner)
```
Also note that your query has a cartesian product:
```sql
SELECT fruit.id AS fruit_id, fruit.type AS fruit_type, fruit.owner_id AS fruit_owner_id, owner.id AS owner_id
FROM fruit, owner
```

so you probably want to do
```py
s = Session(e)
FruitWithPolymorphic = with_polymorphic(Fruit, "*")
WithName = aliased(FruitWithPolymorphic, name='Fruit')
testQuery = s.query(WithName)
testQuery = testQuery.join(WithName.owner).add_entity(Owner)

for r in testQuery.all():
    print(r.Fruit)
    print(r.Owner)
```
notice the `join`. You may also want to use `outerjoin()` instead, since `owner_id` is nullable

**Reply by ske99ey** [reply]
Thanks for suggesting aliased but there is something more going on in our codebase that I won't try to replicate here so we are moving forward using positional access.

In regards to:
> the with_polymorphic construct returns a unnamed object, so you can only access it in a positional way, as first[0].

Just curious why this choice as previously (`models.Fruit.query.with_polymorphic("*")`) a name was attached? We expected it to be the base class / first param given to `with_polymorphic()` to match the previous observed behavior which is what led us to asking here.

**Reply by CaselIT** [reply]
In general v2 is trying to make less assumptions, but here I'm not sure if it's a missing behaviour or intended, cc @zzzeek

**Reply by zzzeek** [reply]
`query.with_polymorphic('*')` only worked for the first entity which was already present, so the name that would be given to the first entity in any case was maintained.  free-floating `with_polymorphic()` is an alias object which can be placed repeatedly, so a name is never assumed for these.

**Reply by CaselIT** [reply]
There doesn't seem to be a way of adding a name directly in the `with_polymorphic` call. Maybe it could be useful to add?

**Reply by zzzeek** [reply]
there's a `name` keyword.  no?

**Reply by zzzeek** [reply]
wow, there's no name keyword.  huh

**Reply by zzzeek** [reply]
well, aliased(with_polymorphic(...), name=name) for now should work

**Reply by zzzeek** [reply]
note this whole time i assumed with_polymoprhic() allowed the name keyword though, so feel free to add that (with tests)

**Reply by CaselIT** [reply]
in case of name we should return `aliased(with_polymorphic(...), name=name)` right?

**Reply by ske99ey** [reply]
> well, aliased(with_polymorphic(...), name=name) for now should work

For unknown reasons, this didn't work for us but I think it has to do with underlying queries we don't fully understand atm. Locally I had added a name param to with_polymorphic() which did work so I would find it useful for our use case at least.

**Reply by CaselIT** [reply]
tracked in https://github.com/sqlalchemy/sqlalchemy/issues/11361

**Reply by CaselIT** [reply]
> in case of name we should return aliased(with_polymorphic(...), name=name) right?

can just pass the name to the aliasedclass constructor

**Comment by ske99ey** [comment, marked as answer]
Thanks all for the discussion - to summarize:
> the with_polymorphic construct returns a unnamed object, so you can only access it in a positional way, as first[0].
To give it a name you can use aliased:

```
s = Session(e)
FruitWithPolymorphic = with_polymorphic(Fruit, "*")
WithName = aliased(FruitWithPolymorphic, name='Fruit')
```

Furthermore, on the 2.0 branch there has been a name field added to `with_polymorphic()` per https://github.com/sqlalchemy/sqlalchemy/issues/11361

**Reply by CaselIT** [reply]
will be released in 2.0.31

**Comment by tribals** [comment]


**Reply by zzzeek** [reply]
"attrs", you mean [attrs](https://www.attrs.org/en/stable/) ?   I dont see how this discussion is related to that, open a new discussion and please provide self-contained scripts in the comment field of the discussion without any dependencies, thanks!
